### ISSUE TITLE
Packaging for more distros / inits?

### ISSUE THREAD

**Issue by endigma** [netbirdio/netbird#363]
Starting to build automated packages for more of the distros in the world would be great, perhaps starting with some of the popular non-debian ones?

Alpine Linux
Artix/Arch
Void Linux
NixOS
Solus

I'm sure given its Go there's a template pkgfile for each of these available if someone were to attempt automating it.

**Comment by endigma** [comment]
Official repositories should do for most distros. If you the first party are committed to maintaining the package and it doesn't have any dependencies I don't foresee you having much of a problem getting updates merged.

As an alternative with less upfront effort you could look into providing consistent versioned artifacts which would make 3rd party packages very simple. Scripts for other inits would also be fairly trivial to include in a doc/ folder or etc.

**Comment by endigma** [comment]
Void Linux

https://github.com/void-linux/void-packages/blob/master/Manual.md

Alpine Linux

https://wiki.alpinelinux.org/wiki/Creating_an_Alpine_package

Arch Linux

https://wiki.archlinux.org/title/creating_packages

NixOS

https://nixos.wiki/wiki/Go

Solus

https://getsol.us/articles/packaging/

**Comment by mlsmaycon** [comment]
Thank you @endigma we will have a look and get back to you soon.

**Comment by hg** [comment]
Consider using Open Build Service instead of reimplementing the wheel. (the code block is to avoid unnecessary linking)

    https://github.com/tonarino/innernet/issues/203#issuecomment-1172847002

**Comment by endigma** [comment]
OBS is great until you realize in reality it only supports:

- openEuler
- openSUSE
- SUSE SLE
- Arch
- Raspbian
- Debian
- Fedora
- ScientificLinux
- RHEL
- CentOS
- Ubuntu
- Univention
- Mageia
- IBM PowerKVM
- AppImage

Which although *technically* covered by AppImage leaves off:

- Solus
- Void
- Alpine
- NixOS
- Gentoo
- Clear Linux

and a lot more, also any BSD

If you don't have the manpower to handle this, non-lock-in solutions like [jordansissel/fpm](https://github.com/jordansissel/fpm) exist, which is somewhat extensible to produce packages for "unsupported" distros.

**Comment by mlsmaycon** [comment]
hello @endigma and @hg, thanks for your suggestions. 

I've checked all the guides and descriptions on how to create the packages. Unfortunately, the project only has 2 maintainers with an extensive roadmap, and we can't move this forward request without community help. 

I've tested the [binary installation](https://netbird.io/docs/getting-started/installation#binary-install) and it worked well with 3/5 of the distros above:

- Alpine
- Artix
- Solus

For Void and NixOS, we had a few issues:

- Void: default init system doesn't have support by our service/init lib https://github.com/kardianos/service, once I am done testing, I will add an example of manual init configuration for it. 
- NixOS: the main problem with the binary installation path is that the /etc is read-only, same as for Void, I will add a configuration.nix example. This will help teams to get started.

With that being said, as we progress with the project and onboard more people we will review this request again in the following quarters.

**Comment by mlsmaycon** [comment]
This is an example of init file for Void:
Create a sv directory:
```shell
mkdir /etc/sv/netbird
```
Add a run file:
```shell
vi /etc/sv/netbird/run
# contents
#!/bin/sh
[ -r conf ] && . ./conf
exec netbird service run --log-level info 1>&2
```
link the file to /var/services
```shell
ln -s /etc/sv/netbird /var/service/
```

**Comment by hg** [comment]
@mlsmaycon personally I wouldn't expect you to support anything but a couple of the most popular distributions. Even major corporations rarely care about anything besides Ubuntu and RHEL.

> OBS is great until you realize in reality it only supports …

… probably 99% of Linux installations out there. The remaining ones mostly cater to enthusiasts who can write their own packages. I really don't think it's reasonable to expect support for Void, Solus, and Clear Linux from a small startup. I don't believe I've ever seen anyone shipping packages for them now that I think of it (or binary packages for any of the BSDs for that matter — you're typically expected to use ports).

Alpine is pretty popular, but it's also frequently updated (and you can easily install 'backports' from the Edge repo, which is perfectly safe for statically linked binaries in my experience).

Note that OBS supports more than they officially claim, although quality may vary (it may not work on less popular architectures). For example, here are projects for cross-compilation onto Windows:

- https://build.opensuse.org/project/show/windows:mingw:win64
- https://build.opensuse.org/project/show/windows:mingw:win32

I'll throw together an OBS repo in the next few days, and we'll see how well it performs. If it works out badly, at least that's one dead end crossed out.

**Comment by misuzu** [comment]
I can help getting netbird to NixOS repository, but I'll need to see existing release scripts first to make a proper package.

**Comment by mlsmaycon** [comment]
awesome, thanks @misuzu for your contribution.

**Comment by Tennyleaz** [comment]
Great if it also support BSD based firewall OS, like OPNSense?
I would love to use it at an gateway.

**Comment by FireMasterK** [comment]
Created an issue in Alpine Linux's aports at https://gitlab.alpinelinux.org/alpine/aports/-/issues/14248

**Comment by FReichelt** [comment]
Is there any update on the BSD support?

**Comment by xromansx** [comment]
I agree BSD wuld be nice to have Opnsense, Pfsense would be perfect for this especially they all support wireguard now.

**Comment by mimi0000oo** [comment]
OPNSense support would be great!

**Comment by nttranbao** [comment]
Please support OpenSense/BSD. Thanks

**Comment by yakaviuk** [comment]
Hi, As for now it is any recommended solution how to install netbird client (not server) on ArchLinux (only CLI, without GUI; `Linux archlinux 6.10.10-arch1-1` )?
I tried `yay` and building using `makepkg -si` but got a lot of errors. 

Thanks in advance

**Comment by mimi0000oo** [comment]
> Hi, As for now it is any recommended solution how to install netbird client (not server) on ArchLinux (only CLI, without GUI; `Linux archlinux 6.10.10-arch1-1` )?
> I tried `yay` and building using `makepkg -si` but got a lot of errors. 
> 
> Thanks in advance

I have netbird installed on arch linux for some time now (cli and gui). I know pacman had an update a few days ago that broke yay stable, but i managed to install `yay-git` and the problem was fixed. Maybe that or some other pacman problem

**Comment by nazarewk** [comment]
Hello @endigma,

We're currently reviewing our open issues and would like to verify if this problem still exists in the [latest NetBird version](https://github.com/netbirdio/netbird/releases).

Could you please confirm if the issue is still there?

We may close this issue temporarily if we don't hear back from you within **2 weeks**, but feel free to reopen it with updated information.

Thanks for your contribution to improving the project!

**Comment by endigma** [comment]
I think this has kinda become a tracking issue for packaging requests? Close it if you like but I bet you'll just get a new one
